src/mongoc/mongoc-async-cmd.c

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright 2014 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <bson.h>
+
+#include "mongoc-client.h"
+#include "mongoc-async-cmd-private.h"
+#include "mongoc-async-private.h"
+#include "mongoc-error.h"
+#include "mongoc-opcode.h"
+#include "mongoc-rpc-private.h"
+#include "mongoc-stream-private.h"
+#include "mongoc-stream-tls.h"
+#include "utlist.h"
+
+#undef MONGOC_LOG_DOMAIN
+#define MONGOC_LOG_DOMAIN "async"
+
+typedef mongoc_async_cmd_result_t (*_monogc_async_cmd_phase_t)(
+   mongoc_async_cmd_t *cmd);
+
+#ifdef MONGOC_ENABLE_SSL
+mongoc_async_cmd_result_t
+_mongoc_async_cmd_phase_tls (mongoc_async_cmd_t *cmd);
+#endif
+mongoc_async_cmd_result_t
+_mongoc_async_cmd_phase_send (mongoc_async_cmd_t *cmd);
+mongoc_async_cmd_result_t
+_mongoc_async_cmd_phase_recv_len (mongoc_async_cmd_t *cmd);
+mongoc_async_cmd_result_t
+_mongoc_async_cmd_phase_recv_rpc (mongoc_async_cmd_t *cmd);
+
+static const _monogc_async_cmd_phase_t gMongocCMDPhases[] = {
+#ifdef MONGOC_ENABLE_SSL
+   _mongoc_async_cmd_phase_tls,
+#endif
+   _mongoc_async_cmd_phase_send,
+   _mongoc_async_cmd_phase_recv_len,
+   _mongoc_async_cmd_phase_recv_rpc,
+};
+
+bool
+mongoc_async_cmd_run (mongoc_async_cmd_t *acmd)
+{
+   mongoc_async_cmd_result_t result;
+   int64_t rtt;
+
+   result = gMongocCMDPhases[acmd->state](acmd);
+
+   rtt = bson_get_monotonic_time () - acmd->start_time;
+
+   if (result == MONGOC_ASYNC_CMD_IN_PROGRESS) {
+      return true;
+   }
+
+   if (result == MONGOC_ASYNC_CMD_SUCCESS) {
+      acmd->cb (result, &acmd->reply, rtt, acmd->data, &acmd->error);
+   } else {
+      /* we're either in ERROR or TIMEOUT */
+      acmd->cb (result, NULL, rtt, acmd->data, &acmd->error);
+   }
+
+   mongoc_async_cmd_destroy (acmd);
+   return false;
+}
+
+void
+_mongoc_async_cmd_init_send (mongoc_async_cmd_t *acmd,
+                             const char         *dbname)
+{
+   bson_snprintf (acmd->ns, sizeof acmd->ns, "%s.$cmd", dbname);
+
+   acmd->rpc.query.msg_len = 0;
+   acmd->rpc.query.request_id = ++acmd->async->request_id;
+   acmd->rpc.query.response_to = 0;
+   acmd->rpc.query.opcode = MONGOC_OPCODE_QUERY;
+   acmd->rpc.query.flags = MONGOC_QUERY_SLAVE_OK;
```

Comment:
This is true.  To simplify/expedite the SDAM work I think making async_cmd a general-purpose tool should be left for another time.  I've opened CDRIVER-548 to make a record of this.